### PR TITLE
Added Classes TLeafF16 and TLeafD32 for truncated floating point types

### PR DESCRIPTION
**Pull request by Triple-S**
Hello,

I implemented the classes TLeafF16 and TLeafD32 which allow to store floating point values using the truncation methods from TBuffer. As the names imply one is meant for the Float16_t type and the other one for the Double32_t type.

The letters to be used in the leaflist are the small letters f and d which sounds reasonable to me since the capital letters F and D are being used for the regular Float_t and Double_t datatypes. Furthermore a range for the datatypes can be specified using the syntax from the TStreamerElement::GetRange() function. To avaid conflicts with the declaration of array branches the range specifier has to be attached to the data type letter instead of the variable name.

I tested the implementation locally and it is working very well. The file size is being significantly reduced using the Float16_t or Double32_t implementation instead of the usual ones. I think this is quite useful since even the precision of Float_t sometimes overshoots the requirements and occupies an unnecessarily large amount of memory.

All the best,
Simon Spies

### CODE REVIEW

File: tree/tree/inc/TLeafF16.h [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+#include "TLeaf.h"
+
+class TLeafF16 : public TLeaf {
+
+protected:
+   Float16_t fMinimum;           ///<  Minimum value if leaf range is specified
+   Float16_t fMaximum;           ///<  Maximum value if leaf range is specified
+   Float16_t *fValue;            ///<! Pointer to data buffer
+   Float16_t **fPointer;         ///<! Address of pointer to data buffer!
+   TStreamerElement *tseFloat16; ///<! StreamerElement used for TBuffer read / write
```
Comment (pcanal) [inline tree/tree/inc/TLeafF16.h]: per coding convention, data member must start with lower case 'f'.

**Comment by Triple-S** [comment]
Hi,

You're right about the naming convention, I totally forgot this and will fix it with the next commit.

Concerning the test, I first have to understand how they work because until now I just executed them.
As far as I know the tests go to the root-project/roottest repository so this requires an independent pull request for this repository, right?

Thanks,
Simon

**Comment by pcanal** [comment]
The test(s) (if small, self contained and leveraging googletest framework) can also go in tree/tree/test.
If in roottest, yes, it needs a separate PR (use the same branch name).

**Comment by Triple-S** [comment]
@pcanal I don't know wheather the system already informed you, but I did the requested changes already 2 days ago.

File: tree/tree/inc/TLeafD32.h [hunk 2]
```diff
@@ -26,11 +26,11 @@
...
-   Double32_t fMinimum;           ///<  Minimum value if leaf range is specified
-   Double32_t fMaximum;           ///<  Maximum value if leaf range is specified
-   Double32_t *fValue;            ///<! Pointer to data buffer
-   Double32_t **fPointer;         ///<! Address of pointer to data buffer
-   TStreamerElement *tseDouble32; ///<! StreamerElement used for TBuffer read / write
+   Double32_t fMinimum;              ///<  Minimum value if leaf range is specified
+   Double32_t fMaximum;              ///<  Maximum value if leaf range is specified
+   Double32_t *fValue;               ///<! Pointer to data buffer
+   Double32_t **fPointer;            ///<! Address of pointer to data buffer
+   TStreamerElement *fStreamingInfo; ///<! StreamerElement used for TBuffer read / write
```
Comment (pcanal) [inline tree/tree/inc/TLeafD32.h]: Please use fElement (a 'StreamerInfo' is a container of SteamerElement).   Thanks.

**Comment by Triple-S** [comment]
@pcanal I just changed the name of the data member. I don't really like fElement, cause it says nothing about its meaning, but OK, I understand why you think fStreamingElement could be confusing.

**Comment by pcanal** [comment]
> cause it says nothing about its meaning,

True.

> fStreamingElement

It is indeed clearer.   I am hesitant because we don't use this wording anywhere else (and have use fElement in TBranchProxy, albeit there it is clearer/more straightforward what the member is about).  fStreamerElement could have work (if it was not just one character away from the type ...)

**Comment by Triple-S** [comment]
> and have use fElement in TBranchProxy

I did not know that it is named fElement there. In this case I think it is really the best solution to use the same wording as there. In the documentation there will anyway be an explanation what the data member is for so I think it is not so problematic.

**Comment by Triple-S** [comment]
> One more thing (sorry), could you also announce this change in README/ReleaseNotes/v618/index.md. Thanks,

You don't have to apologize. I just added a short entry to the release notes summarizing the new functionalities.

File: README/ReleaseNotes/v618/index.md [hunk 3]
```diff
@@ -76,6 +76,14 @@ The methods could be replaced by equivalent methods with other signature:
   - PyROOT: add `AsRNode` helper function to convert RDF nodes to the common RNode type
   - PyROOT: add `AsNumpy` method to export contents of a RDataFrame as a dictionary of numpy arrays
 
+### TLeafF16 and TLeafD32
+  - New leaf classes allowing to store Float and Double values using the truncation methods from TBuffer 
+    (See for example `TBuffer::WriteDouble32`)
+  - The new types can be specified using the type characters 'f' (Float16_t) and 'd' (Double32_t)
```
Comment (pcanal) [inline README/ReleaseNotes/v618/index.md]: Consider adding usage example.  Thanks.

**Comment by Triple-S** [comment]
Just added a short example to the release notes showing how to create branches of the new types with and without specifying a range or amount of bits.

**Comment by pcanal** [comment]
Perfect. Thanks.

**Pull request merged**